### PR TITLE
[9825] Updating Fast-DDS to c++20

### PR DESCRIPTION
**Pull request by MiguelBarro**
This pull request addresses https://github.com/eProsima/Fast-DDS/issues/1381
Henceforth we can build on C++ 20 by doing:
```bash
$ colcon build --cmake-args -DFORCE_CXX:string=20
```
tested on clang 11

### CODE REVIEW

File: cmake/modules/CheckLibatomic.cmake [hunk 1]
```diff
@@ -5,13 +5,24 @@
...
+            uint32_t write_p;
+            uint32_t free_cells;
+        };
+
+        Pointer dummy;
+        std::atomic<Pointer> pointer;
+        pointer = dummy;
+
+return 0;
+}"
```
Comment (MiguelCompany) [inline cmake/modules/CheckLibatomic.cmake:25]: Please indent this block

File: include/fastrtps/types/DynamicTypeBuilderPtr.h [hunk 2]
```diff
@@ -42,6 +42,16 @@ class DynamicTypeBuilder_ptr : public std::shared_ptr<DynamicTypeBuilder>
...
+
+    RTPS_DllAPI bool operator!=(std::nullptr_t) const
+    {
+            return bool(*this);
+    }
+
+    RTPS_DllAPI bool operator==(std::nullptr_t) const
+    {
+            return !*this;
+    }
```
Comment (MiguelCompany) [inline include/fastrtps/types/DynamicTypeBuilderPtr.h]: Adding new API should add corresponding tests.

File: include/fastrtps/types/DynamicTypePtr.h [hunk 3]
```diff
@@ -54,6 +54,16 @@ class DynamicType_ptr : public std::shared_ptr<DynamicType>
...
+
+    RTPS_DllAPI bool operator!=(std::nullptr_t) const
+    {
+            return bool(*this);
+    }
+
+    RTPS_DllAPI bool operator==(std::nullptr_t) const
+    {
+            return !*this;
+    }
```
Comment (MiguelCompany) [inline include/fastrtps/types/DynamicTypePtr.h]: Adding new API should add corresponding tests.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3289/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/500/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2936/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3321/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3295/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/506/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2942/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3327/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3299/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/510/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2946/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3331/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3303/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/514/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2950/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3335/)

File: cmake/modules/CheckLibatomic.cmake (hunk 1 above)
Comment (MiguelBarro) [reply cmake/modules/CheckLibatomic.cmake]: 👍

File: include/fastrtps/types/DynamicTypeBuilderPtr.h (hunk 2 above)
Comment (MiguelBarro) [reply include/fastrtps/types/DynamicTypeBuilderPtr.h]: Already checked in \test\unittest\dynamic_types\DynamicTypesTests.cpp

File: include/fastrtps/types/DynamicTypePtr.h (hunk 3 above)
Comment (MiguelBarro) [reply include/fastrtps/types/DynamicTypePtr.h]: Already checked in \test\unittest\dynamic_types\DynamicTypesTests.cpp

**Review by MiguelCompany — DISMISSED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3308/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/519/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2956/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3340/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3353/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/558/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3004/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3385/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3375/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/575/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3027/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3407/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3376/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/576/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3028/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3408/)

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**